demo_apps/llama-on-prem.md

```diff
@@ -0,0 +1,180 @@
+# Llama 2 On-Prem Inference Using vLLM and TGI
+
+Enterprise customers may prefer to deploy Llama 2 on-prem and run Llama in their own servers. This tutorial shows how to use Llama 2 with [vLLM](https://github.com/vllm-project/vllm) and Hugging Face [TGI](https://github.com/huggingface/text-generation-inference), two leading open-source tools to deploy and serve LLMs, and how to create vLLM and TGI hosted Llama 2 instances with [LangChain](https://www.langchain.com/), an open-source LLM app development framework which we used for our earlier demo apps with Llama 2 running on [local Mac](https://github.com/facebookresearch/llama-recipes/blob/main/demo_apps/HelloLlamaLocal.ipynb) or [Replicate cloud](https://github.com/facebookresearch/llama-recipes/blob/main/demo_apps/HelloLlamaCloud.ipynb). See [here](https://medium.com/@rohit.k/tgi-vs-vllm-making-informed-choices-for-llm-deployment-37c56d7ff705) for a detailed comparison of vLLM and TGI.
+
+We'll use the Amazon EC2 instance running Ubuntu with an A10G 24GB GPU as an example of running vLLM and TGI with Llama 2, and you can replace this with your own server to implement on-prem Llama 2 deployment.
+
+The Colab notebook to connect via LangChain with Llama 2 hosted as the vLLM and TGI API services is [here](https://colab.research.google.com/drive/1rYWLdgTGIU1yCHmRpAOB2D-84fPzmOJg?usp=sharing), also shown in the sections below.
+
+This tutorial assumes that you you have been granted access to the Meta Llama 2 on Hugging Face - you can open a Hugging Face Meta model page [here](https://huggingface.co/meta-llama/Llama-2-13b-chat-hf) to confirm that you see "Gated model You have been granted access to this model"; if you don't see the "granted access" message, simply follow the instructions under "Access Llama 2 on Hugging Face" in the page. 
+
+You'll also need your Hugging Face access token which you can get at your Settings page [here](https://huggingface.co/settings/tokens).
+
+## Setting up vLLM with Llama 2
+
+On a terminal, run the following commands:
+```
+conda create -n vllm python=3.8
+conda activate vllm
+pip install vllm
+cd <your_work_folder>
+git clone https://github.com/vllm-project/vllm
+cd vllm/vllm/entrypoints/
+```
+
+Then run `huggingface-cli login` and copy and paste your Hugging Face access token to complete the login.
+
+There are two ways to deploy Llama 2 via vLLM, as a general API server or an OpenAI-compatible server.
+
+### Deploying Llama 2 as an API Server
+
+Run the command below to deploy vLLM as a general Llama 2 service:
+
+```
+python api_server.py --host 0.0.0.0 --port 5000 --model meta-llama/Llama-2-7b-chat-hf
+```
+
+Then on another terminal you can run:
+
+```
+curl http://localhost:5000/generate -d '{
```

Comment:
Besides the CLI, should we provide a python example snippet as well?

Reply:
The Google Colab shows complete Python code to talk to the servers. The curl commands are just for quick verification.

Reply:
OK, I did see in the colab we inferenced the API server via Langchain. I wonder if we want to add some example using POST in python, which is more raw. I have some example, let me link it later.

Reply:
The reason is that vllm documentation is not so clear about this part. If we can add some examples here, I think it would helpful

Reply:
Here is an [example](https://github.com/WuhanMonkey/Azure_API_hosting_benchmark/blob/main/vm-local/70b_chat_vllm_benchmark.py#L148) how I used the urlib to post requests to the api server.

Reply:
Thanks but your examples seems to be in your private repo so I can't open it. 

Converting the curl command to a Python script is really straightforward, that's why I had the content "Now in your Llama client app, you can make an HTTP request as the `curl` command above to send a query to Llama and parse the response." I assume if one needs that script, he/she can simply ask ChatGPT like this :)
https://chat.openai.com/c/013cc516-71db-4ee7-b8e6-3ec946734cfe

Reply:
Also, I only showed the creation of the Llama instance "llm" in the Colab - but using LangChain or LlamaIndex will quickly be a lot more powerful and effective, when building Llama RAG apps, than using the raw post request which is mainly for quick API test.

Reply:
Can't access your ChatGPT result :) it doesn't load for me. I have invited you to that repo, so you can take a look. Sure. I think it is a different use case, especially when you are running benchmarks and doesn't want to use any framework. Might not a good fit in this PR.

Reply:
Can you access the link now? https://chat.openai.com/share/d289e3f8-9d84-4eea-b4d0-fafd752150c9 (I should've clicked the Share button first). If you still can't, you can just enter in ChatGPT this:

write a python script that implements the curl command below:

curl http://localhost:5000/generate -d '{
        "prompt": "Who wrote the book Innovators dilemma?",
        "max_tokens": 300,
        "temperature": 0
    }'